network/p2p/peerstore/peerstore.go

```diff
@@ -205,19 +205,29 @@ func (ps *PeerStore) UpdateConnectionTime(addrOrPeerID string, provisionalTime t
 }
 
 // ReplacePeerList replaces the peer list for the given networkName and role.
-func (ps *PeerStore) ReplacePeerList(addressesThey []*peer.AddrInfo, networkName string, role phonebook.PhoneBookEntryRoles) {
+func (ps *PeerStore) ReplacePeerList(addressesThey []*peer.AddrInfo, networkName string, role phonebook.Roles) {
 	// prepare a map of items we'd like to remove.
 	removeItems := make(map[peer.ID]bool, 0)
 	peerIDs := ps.Peers()
 	for _, pid := range peerIDs {
 		data, _ := ps.Get(pid, addressDataKey)
 		if data != nil {
 			ad := data.(addressData)
+			updated := false
 			ad.mu.RLock()
```

Comment:
This is locking access to `networkNames` right? Wondering about the remove on line 222 being under read lock.

Reply:
it protects `ad.networkNames` where `ad` is a copy (so that roles access is safe) and `networkNames` is a map => a shallow copy and access needs to be synchronized.

Reply:
to clarify, `ad` is a local variable

```golang
data, _ := ps.Get(pid, addressDataKey)
ad := data.(addressData)
```
and its `ad.role` member is updated without a lock because of that. Later `ad` is saved back to peerstore:

```golang
_ = ps.Put(pid, addressDataKey, ad)
```